src/3.md

```diff
@@ -150,27 +181,159 @@ impl<T> Point<T> {
+    }
+}
+```
+
+The type can then be used as normal:
+
+```rust
+#trait Printable {
+#    fn format(&self) -> String;
+#}
+#struct Person { first_name: String, last_name: String }
+#
+#impl Printable for Person {
+#    fn format(&self) -> String {
+#        format!("{} {}", self.first_name, self.last_name)
+#    }
+#}
+#
+
+fn main() {
+    let named = Person {
+        first_name: String::from("Joris"),
+        last_name: String::from("Bohnson"),
+    };
+    println!("{}", named.format());
+}
+```
+Traits can be used to restrict the type of a generic, allowing use of any type that implements the trait: `f<T: Trait1 + Trait2>(a: T, b: T)`:
+
+```rust
+fn bold_fmt<T : Printable>(original: T) -> String {
+    format!("**{}**", original.format())
+}
+```
+
+This pattern of a `.format` seems awfully useful, and luckily enough Rust provides a `Display` trait to print a struct. `println!()` calls `fmt` automatically on any `Display` types when using `{}` as the placeholder:
 
 ```rust, editable
-pub trait Summary {
-    fn summarise(&self) -> String;
+use std::fmt;
+
+struct Person { first_name: String, last_name: String }
+
+impl fmt::Display for Person {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} {}", self.first_name, self.last_name)
+    }
+}
+
+fn main() {
+    let named = Person {
+        first_name: String::from("Joris"),
+        last_name: String::from("Bohnson"),
+    };
+    println!("{}", named);
 }
 ```
 
-Other types can then implement traits as shown.
+`Display` is meant for user-facing representation of a type, but there also exists `Debug` that is intended to fully describe a type instance to the programmer (like `repr` in Python). This uses the `{:?}` placeholder.
 
-It is often useful to provide default implementations for methods of a trait, that then types can choose to override if they wish, or use the default behaviour.
+Writing these, especially `Debug`, can get very tedious, so Rust allows you to derive some traits, including `Debug`, `PartialEq`, and `Hash`. This is just like Haskell's `deriving`.
 
-Traits are used to specify bounds on generics and
+```rust, editable
+#[derive(Debug, PartialEq)]
+struct Person { 
+    first_name: String, 
+    last_name: String 
+}
 
-Traits and generics form the backbone of the type system, and theres an awful lot you can do with them. A few examples, in descending order of "how ~~cursed~~ fancy is this:"
+fn main() {
+    let named = Person {
+        first_name: String::from("Joris"),
+        last_name: String::from("Bohnson"),
+    };
+    
+    let named2 = Person {
+        first_name: String::from("Joris"),
+        last_name: String::from("Bohnson"),
+    };
+
+    println!("{:?} {}", named, named == named2);
+}
+```
+
+We have already seen Rust must use `Box`es for heap allocated memory. Rust's compiler needs to know the size of a function's return type, but traits aren't fixed sizes. To avoid this, return types need to be boxed:
```

Comment:
trait *objects* have no fixed size, traits themselves are an abstract thing and don't have any size